aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -96,206 +127,442 @@ impl AggregatorID {
+                EEXPECTED_UNDERFLOW,
+            ));
+        }
+    }
+    Ok(())
+}
+
+/// Internal aggregator data structure.
+#[derive(Debug)]
+pub struct Aggregator {
+    // The identifier used to identify the aggregator.
+    id: AggregatorID,
+    // Describes an upper bound of an aggregator. If value of the aggregator
+    // exceeds it, the aggregator overflows.
     // TODO: Currently this is a single u128 value since we use 0 as a trivial
     // lower bound. If we want to support custom lower bounds, or have more
     // complex postconditions, we should factor this out in its own struct.
-    limit: u128,
-    // Describes values seen by this aggregator. Note that if aggregator knows
-    // its value, then storing history doesn't make sense.
-    history: Option<History>,
+    max_value: u128,
+    // Describes a state of an aggregator.
+    state: AggregatorState,
 }
 
 impl Aggregator {
-    /// Records observed delta in history. Should be called after an operation
-    /// to record its side-effects.
-    fn record(&mut self) {
-        if let Some(history) = self.history.as_mut() {
-            match self.state {
-                AggregatorState::PositiveDelta => history.record_positive(self.value),
-                AggregatorState::NegativeDelta => history.record_negative(self.value),
-                AggregatorState::Data => {
-                    unreachable!("history is not tracked when aggregator knows its value")
-                },
-            }
+    fn get_mut_history(&mut self) -> Option<&mut DeltaHistory> {
+        if let AggregatorState::Delta { history, .. } = &mut self.state {
+            return Some(history);
         }
+        None
+    }
+
+    pub fn is_data(&self) -> bool {
+        matches!(self.state, AggregatorState::Data { .. })
+    }
+
+    pub fn is_delta(&self) -> bool {
+        matches!(self.state, AggregatorState::Delta { .. })
     }
 
-    /// Validates if aggregator's history is correct when applied to
-    /// the `base_value`. For example, if history observed a delta of
-    /// +100, and the aggregator limit is 150, then the base value of
-    /// 60 will not pass validation (60 + 100 > 150), but the base value
-    /// of 30 will (30 + 100 < 150).
-    fn validate_history(&self, base_value: u128) -> PartialVMResult<()> {
-        let history = self
-            .history
-            .as_ref()
-            .expect("History should be set for validation");
+    #[allow(dead_code)]
+    fn get_value(&self) -> BoundedMathResult<u128> {
+        match self.state {
+            AggregatorState::Data { value } => Ok(value),
+            AggregatorState::Delta {
+                speculative_start_value,
+                delta,
+                history: _,
+            } => addition_deltavalue(speculative_start_value.get_value(), delta, self.max_value),
```

Comment:
connecting to the conversation re: (.get value of speculative value and read snapshot example) - .get_value here should probably not be able to return "last committed" value, as that is not registered as a read and validated.